### PR TITLE
Fix compiling async arrows in uncompiled class fields

### LINKED ISSUE #14749: [Bug]: _this is not defined from asyncToGenerator.js
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Other (Next.js, Gatsby, vue-cli, ...)

### Input code

```
updateDiscussionTitle = async (): Promise<void> => {
    await this.legacyWorkspaceService.eznoteData.updateDiscussionTitle(this.data.discussion.discussionId, this.newTitle);
    this.data.discussion.title = this.newTitle;
    this.dialogRef.close();
  };
```

```
updateDiscussionTitle = (0,node_modules_babel_runtime_helpers_esm_asyncToGenerator_js__WEBPACK_IMPORTED_MODULE_0__["default"])(function* () {
    yield _this.legacyWorkspaceService.eznoteData.updateDiscussionTitle(_this.data.discussion.discussionId, _this.newTitle);
    _this.data.discussion.title = _this.newTitle;

    _this.dialogRef.close();
  });
```

### Configuration file name

_No response_

### Configuration

_No response_

### Current and expected behavior

_this is not being created in this instance. If i async before the function name and don't arrow it's fine. If i IIFE the function in a weird way it works.

I recently Updated to Angular 14, and changed the target in the tsconfig to 2022 from 2020. If i revert back to 2020 it works as well.  I have been searching all day for an answer and am hoping this is related to y'all at this point since i can trick it with other ways of formatting the function.

### Environment

System:
    OS: Windows 10 10.0.19043
  Binaries:
    Node: 16.13.2 - C:\Program Files\nodejs\node.EXE
    npm: 8.1.2 - C:\Program Files\nodejs\npm.CMD
  npmPackages: (from build-angular)
    "@babel/core": "7.17.10",
        "@babel/generator": "7.17.10",
        "@babel/plugin-proposal-async-generator-functions": "7.16.8",
        "@babel/plugin-transform-async-to-generator": "7.16.8",
        "typescript": 4.7.3,
        "angular": 14.0.5

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #14749
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR should be reviewed commit-by-commit: the first one is a refactor to help TS inference understand the types of different variables (we had three different logic paths with different types all mixed together); the second one is the bug fix.

The generated output for async functions in uncompiled class fields is ugly, but it's uncommon enough that it should be fine (but not _too uncommon_, because Angular forces async functions compilation).

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14752"></a>

### CODE REVIEW

**Review by liuxingbaoyu — APPROVED** [review]

File: packages/babel-helper-wrap-function/src/index.ts [hunk 1]
```diff
@@ -95,19 +94,21 @@ function plainFunction(
...
-  const { node } = path;
-  const isDeclaration = isFunctionDeclaration(node);
-  const isArrow = isArrowFunctionExpression(node);
-
   let functionId = null;
-  if (isArrow) {
-    (path as NodePath<typeof node>).arrowFunctionToExpression({ noNewArrows });
+  let node;
+  if (path.isArrowFunctionExpression()) {
+    path = path.arrowFunctionToExpression({ noNewArrows });
```
Comment (JLHwung) [inline packages/babel-helper-wrap-function/src/index.ts:100]: `arrowFunctionToExpression` returns `void`, I am surprised that this works.

File: packages/babel-traverse/src/path/conversion.ts [hunk 2]
```diff
@@ -151,7 +158,7 @@ export function arrowFunctionToExpression(
     specCompliant?: boolean | void;
     noNewArrows?: boolean;
   } = {},
-) {
+): NodePath<Exclude<t.Function, t.Method | t.ArrowFunctionExpression>> {
```
Comment (JLHwung) [inline packages/babel-traverse/src/path/conversion.ts:161]: `path.arrowFunctionToExpression` is an API, should we defer the changes to minor?

File: packages/babel-helper-wrap-function/src/index.ts [hunk 3]
```diff
@@ -95,19 +94,21 @@ function plainFunction(
...
+    node = path.node as t.FunctionDeclaration | t.FunctionExpression;
```
[the remark this answers is not in the document]
Comment (nicolo-ribaudo) [reply packages/babel-helper-wrap-function/src/index.ts:101]: No, it still includes `t.ArrowFunctionExpression` :confused:

[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-wrap-function/src/index.ts:101]: Ah it is likely confused by the reassignment.

File: packages/babel-traverse/src/path/conversion.ts (hunk 2 above)
Comment (nicolo-ribaudo) [reply packages/babel-traverse/src/path/conversion.ts:161]: I think we have modified these internal plugin APIs in the past even in patch releases, and this change isn't adding "new functionality" but just exposing more info from an existing functionality.

However, if you prefer to wait that's fine

**Review by JLHwung — APPROVED** [review]
The return value of `arrowFunctionToExpression` is practical. If someone reported that their workflow is broken by this PR, we can reland it in the minor.

File: packages/babel-helper-wrap-function/src/index.ts [hunk 4]
```diff
@@ -6,19 +6,39 @@ import {
   callExpression,
   functionExpression,
   isAssignmentPattern,
+  isFunctionDeclaration,
   isRestElement,
   returnStatement,
 } from "@babel/types";
 import type * as t from "@babel/types";
 
+type ExpressionWrapperBuilder<ExtraBody extends t.Node[]> = (
```
Comment (tdnumerique) [inline packages/babel-helper-wrap-function/src/index.ts:15]: I have never work on a project where we needed such a custom type. It leave me suspicious.

**Comment by Fre4ked** [comment]
> The return value of `arrowFunctionToExpression` is practical. If someone reported that their workflow is broken by this PR, we can reland it in the minor.

I can confirm that this breaks our build. If using npx-quill (and possibly many other depedencies), the `.arrowFunctionToExpression({...})` returns `undefined`. Later on this gives a `\node_modules\ngx-quill\fesm2020\ngx-quill.mjs: Cannot read property 'node' of undefined` and thus the build breaks.

**Comment by JLHwung** [comment]
@Fre4ked Just to clarify:

Before this PR `NodePath#arrowFunctionToExpression` always returns `undefined`.
After this PR `NodePath#arrowFunctionToExpression` always returns a `NodePath<FunctionExpression>`. 

Based the provided error message, I can't tell how the new behaviour will break your build. Can you open a new issue and provide a reproduction repo?

**Comment by Fre4ked** [comment]
The error happens when `path.node` is accessed in line 101. Previously the undefined result wasn't accessed later on. In the new version, the result of calling `arrowFunctionToExpression` is saved as `path`, overwriting it, meaning `path` is now undefined and accessing the `node` property throws an exception.

**Comment by nicolo-ribaudo** [comment]
This is a compatibility issue between new `@babel/helper-wrap-function` versions and old `@babel/traverse` versions. As a workaround, update `@babel/traverse` (you might need to delete it from your lockfile and re-install your dependencies)

**Pull request merged**